ui/tests/unit/services/api-test.js

```diff
@@ -0,0 +1,202 @@
+
+    const authService = this.owner.lookup('service:auth');
+    this.setLastFetch = sinon.spy(authService, 'setLastFetch');
+    this.currentToken = sinon.stub(authService, 'currentToken').value('foobar');
+
+    const namespaceService = this.owner.lookup('service:namespace');
+    this.namespace = sinon.stub(namespaceService, 'path').value('another-ns');
+
+    const controlGroupService = this.owner.lookup('service:control-group');
+    this.wrapInfo = { token: 'ctrl-group', accessor: '84tfdfd5pQ5vOOEMxC2o3Ymt' };
+    this.tokenForUrl = sinon.stub(controlGroupService, 'tokenForUrl').returns(this.wrapInfo);
+    this.deleteControlGroupToken = sinon.spy(controlGroupService, 'deleteControlGroupToken');
+
+    const flashMessageService = this.owner.lookup('service:flash-messages');
+    this.info = sinon.spy(flashMessageService, 'info');
+
+    this.url = '/v1/sys/capabilities-self';
+  });
+
+  test('it should set last fetch time', async function (assert) {
+    await this.apiService.setLastFetch({ url: '/v1/sys/health' });
+    assert.true(this.setLastFetch.notCalled, 'Last fetch is not set for polling url');
+
+    await this.apiService.setLastFetch({ url: '/v1/auth/token/lookup-self' });
+    assert.true(this.setLastFetch.calledOnce, 'Last fetch is set for non polling url');
+  });
+
+  test('it should get control group token', async function (assert) {
+    const context = {
+      url: this.url,
+      init: {
+        method: 'GET',
+        headers: { 'X-Vault-Token': 'root' },
+      },
+    };
+
+    this.tokenForUrl.returns(undefined);
+    const noTokenContext = await this.apiService.getControlGroupToken(context);
+
+    assert.true(this.tokenForUrl.calledWith(context.url), 'Url is passed to tokenForUrl method');
+    assert.deepEqual(context, noTokenContext, 'Original context is returned when no token is present');
+
+    this.tokenForUrl.returns(this.wrapInfo);
+
+    const { token } = this.wrapInfo;
+    const tokenContext = await this.apiService.getControlGroupToken(context);
+    const newContext = {
+      url: '/v1/sys/wrapping/unwrap',
+      init: {
+        method: 'POST',
+        headers: { 'X-Vault-Token': token },
+        body: JSON.stringify({ token }),
+      },
+    };
+
+    assert.deepEqual(tokenContext, newContext, 'New context is returned when token is present');
+  });
+
+  test('it should set headers', async function (assert) {
+    const headers = {
+      'X-Vault-Token': 'root',
+      'X-Vault-Namespace': 'ns1',
+    };
+
+    const {
+      init: { headers: originalHeaders },
+    } = await this.apiService.setHeaders({ init: { headers } });
+
+    assert.strictEqual(
+      originalHeaders.get('X-Vault-Token'),
+      'root',
+      'Original token header value is preserved'
+    );
+    assert.strictEqual(
+      originalHeaders.get('X-Vault-Namespace'),
+      'ns1',
+      'Original namespace header value is preserved'
+    );
+
+    const {
+      init: { headers: newHeaders },
```

Comment:
i may just be following this wrong, but should these variable names be switched (and this called `originalHeaders`) since these are set in the `beforeEach` hook and then here were checking that if nothing is in the init object, the headers are unchanged? And above `init` has "new headers" and we assert those are returned?

Reply:
> since these are set in the beforeEach hook

Not sure what you are referring to here but this test steps through the various states of the `setHeaders` method. I named the headers returned from the first call `originalHeaders` since the intention is that the original header values (defined in the headers const on line 74 and passed into the `setHeaders` function) are preserved and not overriden in the `setHeaders` method. I then went with `newHeaders` since I passed in an empty object and was expecting to get back headers for token and namespace.

Open to naming this differently if it's confusing or using `let` instead and just accessing them like `init.headers`? Maybe that is better?

Reply:
by "these are set" i meant the namespace and token values set in the `beforeEach` hook. To me those are the "original" headers because they're set by the test context. Which would make the headers defined on 74 "overriding" those initial values. But I also think I don't fully understand the purpose of the `init` object. Maybe it'll be easier to parse seeing it live in action.